oximeter/instruments/src/kstat/sampler.rs

```diff
@@ -513,149 +443,261 @@ impl KstatSamplerWorker {
+                        sample_timeouts.push(next_timeout);
+                    }
+                }
+            }
+        }
+    }
 
+    // Handle a message on the worker's inbox.
+    fn handle_inbox_request(
+        &mut self,
+        request: Request,
+    ) -> Option<YieldIdAfter> {
+        match request {
+            Request::AddTarget { target, details, reply_tx } => {
+                match self.add_target(target, details) {
+                    Ok(id) => {
+                        trace!(
+                            self.log,
+                            "added target with timeout";
+                            "id" => ?id,
+                            "details" => ?details,
+                        );
+                        match reply_tx.send(Ok(id)) {
+                            Ok(_) => trace!(self.log, "sent reply"),
+                            Err(e) => error!(
+                                self.log,
+                                "failed to send reply";
+                                "id" => ?id,
+                                "error" => ?e,
+                            ),
                         }
-                        Request::RemoveTarget { id, reply_tx } => {
-                            self.targets.remove(&id);
-                            if let Some(remaining_samples) = self.samples.lock().unwrap().remove(&id) {
-                                if !remaining_samples.is_empty() {
-                                    warn!(
-                                        self.log,
-                                        "target removed with queued samples";
-                                        "id" => ?id,
-                                        "n_samples" => remaining_samples.len(),
-                                    );
-                                }
-                            }
-                            match reply_tx.send(Ok(())) {
-                                Ok(_) => trace!(self.log, "sent reply"),
-                                Err(e) => error!(
-                                    self.log,
-                                    "failed to send reply";
-                                    "error" => ?e,
-                                )
-                            }
+                        Some(YieldIdAfter::new(id, details.interval))
+                    }
+                    Err(e) => {
+                        error!(
+                            self.log,
+                            "failed to add target";
+                            "error" => ?e,
+                        );
+                        match reply_tx.send(Err(e)) {
+                            Ok(_) => trace!(self.log, "sent reply"),
+                            Err(e) => error!(
+                                self.log,
+                                "failed to send reply";
+                                "error" => ?e,
+                            ),
                         }
```

Comment:
i wondered briefly if we could make this (and the similar code in other match arms) a bit less repetitive by having the `match` on whether the operation succeeds just return the reply, and then send that to `reply_tx` in one place, rather than in two. but, we can't easily do that because we also decide whether or not to return `Some(YieldIdAfter::new(...))` based on whether or not the operation succeeds. Perhaps we could make this a bit less repetitive if we pulled the code for sending a message to `reply_tx` and logging if that fails into a function, but it's not really a big deal either way...